src/ApiClient.js

```diff
@@ -0,0 +1,71 @@
+// This class is a reference implementation. If you desire additional
+// functionality, you can work with the config options here, or copy this code
+// into your project and customize the ApiClient to your needs.
+
+class ApiClient {
+  constructor(passedConfig) {
+    const defaultConfig = {
+      methods: ['get', 'post', 'put', 'patch', 'delete'],
+      credentials: 'same-origin',
+      defaultHeaders: {
+        'Accept': 'application/json',
+        'Content-Type': 'application/json'
+      },
+      format: 'json' // json, blob, text, arrayBuffer, ...
+    }
+
+    const config = {
+      ...defaultConfig,
+      ...passedConfig
+    }
+
+    if (!config.basePath) {
+      // e.g. 'https://example.com/api/v3'
+      throw new Error('You must pass a base path to the ApiClient')
+    }
+
+    config.methods.forEach(method => {
+      this[method] = (path, { params, data, otherConfig } = {}) => {
+        const queryString = this.queryString(params)
+        const { basePath, defaultHeaders, format, ...fetchConfig } = config
+        const headers = {
+          this.getHeaders(defaultHeaders, data),
+          ...fetchConfig.headers
+        }
+
+        return new Promise((resolve, reject) => {
+          fetch(basePath + path + this.queryString(), {
+            ...fetchConfig,
+            ...otherConfig,
```

Comment:
Yeah, I was thinking this morning this could get crazy, and wondering the best way to make it work. It took a while to get around to the idea of passing otherConfig AND passing passedConfig to the constructor. I'll take a stab at refactoring when I next sit down with this.